### PR TITLE
Get with field index from pattern slice instead of directly indexing

### LINKED ISSUE #82772: Trying to pattern-match field `1` of `Box` causes an ICE
### Code

```Rust
fn main() {
    let Box { 1: _, .. }: Box<()>;
}
```

[[playground]](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=44d88eb3c3d1860eec0c9d8f50550592)

### Meta

`rustc --version --verbose`:
```
1.52.0-nightly

(2021-03-01 4f20caa6258d4c74ce6b)
```

### Error output

```
thread 'rustc' panicked at 'index out of bounds: the len is 1 but the index is 1', compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:1346:21
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4f20caa6258d4c74ce6b316fd347e3efe81cf557/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/4f20caa6258d4c74ce6b316fd347e3efe81cf557/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/4f20caa6258d4c74ce6b316fd347e3efe81cf557/library/core/src/panicking.rs:69:5
   3: rustc_mir_build::thir::pattern::deconstruct_pat::Fields::replace_with_pattern_arguments
   4: rustc_mir_build::thir::pattern::usefulness::PatStack::pop_head_constructor
   5: rustc_mir_build::thir::pattern::usefulness::is_useful
   6: <core::iter::adapters::copied::Copied<I> as core::iter::traits::iterator::Iterator>::fold
   7: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
   8: rustc_mir_build::thir::pattern::usefulness::compute_match_usefulness
   9: rustc_mir_build::thir::pattern::check_match::MatchVisitor::check_irrefutable
  10: <rustc_mir_build::thir::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_local
  11: rustc_hir::intravisit::walk_block
  12: <rustc_mir_build::thir::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr
  13: rustc_mir_build::thir::pattern::check_match::check_match
  14: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  15: rustc_data_structures::stack::ensure_sufficient_stack
  16: rustc_query_system::query::plumbing::force_query_with_job
  17: rustc_query_system::query::plumbing::get_query_impl
  18: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::check_match
  19: rustc_middle::ty::<impl rustc_middle::ty::context::TyCtxt>::par_body_owners
  20: rustc_session::utils::<impl rustc_session::session::Session>::time
  21: rustc_interface::passes::analysis
  22: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  23: rustc_data_structures::stack::ensure_sufficient_stack
  24: rustc_query_system::query::plumbing::force_query_with_job
  25: rustc_query_system::query::plumbing::get_query_impl
  26: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  27: rustc_interface::passes::QueryContext::enter
  28: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  29: rustc_span::with_source_map
  30: rustc_interface::interface::create_compiler_and_run
  31: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

query stack during panic:
#0 [check_match] match-checking `main`
#1 [analysis] running analysis passes on this crate
end of query stack
```

</p>
</details>

### PR DESCRIPTION
**Pull request by csmoe**
Closes #82772 
r? @estebank 

https://github.com/rust-lang/rust/pull/82789#issuecomment-796921977
> @estebank So the real cause is we only generate single pattern for Box here
https://github.com/csmoe/rust/blob/615b03aeaa8ce9819de7828740ab3cd7def4fa76/compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs#L1130-L1132
But in the replacing function, it tries to index on the 1-length pattern slice with field 1, thus out of bounds.
https://github.com/csmoe/rust/blob/615b03aeaa8ce9819de7828740ab3cd7def4fa76/compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs#L1346

### CODE REVIEW

File: compiler/rustc_typeck/src/check/pat.rs [hunk 1]
```diff
@@ -1168,7 +1168,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             .fields
             .iter()
             .enumerate()
-            .map(|(i, field)| (field.ident.normalize_to_macros_2_0(), (i, field)))
+            .filter_map(|(i, field)| {
+                if field.vis.is_visible_locally() {
```
Comment (petrochenkov) [inline compiler/rustc_typeck/src/check/pat.rs:1172]: This is *very* suspicious and is unlikely to be correct in presence on macro hygiene (unless this code only affects some unimportant like diagnostic details and can use heuristics).

Comment (csmoe) [reply compiler/rustc_typeck/src/check/pat.rs]: yep, this check is only used to filter out invisiable fileds in error reporting.

File: compiler/rustc_typeck/src/check/pat.rs [hunk 2]
```diff
@@ -1359,6 +1370,41 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        .join(", ")
+                ),
+                "are",
+            )
+        };
+        let err = struct_span_err!(
+            self.tcx.sess,
+            spans,
+            E0603,
+            "cannot match on {} of {} `{}`, which {} invisible in current scope",
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/pat.rs:1399]:
```suggestion
            "cannot match on {} of {} `{}`, which {} not accessible in current scope",
```

**Comment by estebank** [comment]
Could you add a test for the more than one field case?

File: compiler/rustc_typeck/src/check/pat.rs (hunk 1 above)
Comment (csmoe) [reply compiler/rustc_typeck/src/check/pat.rs]: my bad, the previous single `is_visible_locally` didn't account the private items inside local crate.
Fixed now I think? (check the visibility of field in current pat's DefTree)

**Comment by estebank** [comment]
I have a theory: we already emit privacy errors pointing at the ADT on some cases of private fields, so the error for the field themselves would only be redundant. The ICE you're fixing _only_ happens on non-existing fields, and the new error is just precluding getting to that. Could you verify whether my theory is correct? I'm swamped until next week to do this kind of light digging :)

**Comment by csmoe** [comment]
@estebank Precisely this ice is triggered by non-accessible fields, as field `1` of Box actually exists but is private. So what I'm trying to do here is filtering out non-accessible field from non-existent error checking.

**Comment by csmoe** [comment]
@estebank So the real cause is we only generate single pattern for Box here
https://github.com/csmoe/rust/blob/615b03aeaa8ce9819de7828740ab3cd7def4fa76/compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs#L1130-L1132
But in the replacing function, it tries to index on the 1-length pattern slice with field index 1, thus out of bounds.
https://github.com/csmoe/rust/blob/615b03aeaa8ce9819de7828740ab3cd7def4fa76/compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs#L1346

**Pull request merged**